File: app/src/main/java/app/fynnjason/fivesingletondemo/java/LazyLoadSingleton.java
package app.fynnjason.fivesingletondemo.java;

/**
 * Created by FynnJason on 2017/12/25.
 * 大家熟知的懒加载方式(线程非安全)，也叫懒汉式
 */

public class LazyLoadSingleton {
    //第一步：私有构造
    private LazyLoadSingleton() {

    }
    //第二步：创建私有构造对象
    private static LazyLoadSingleton INSTANCE;

    //第二步：提供对外获取方法
    public static LazyLoadSingleton getInstance() {
        if (INSTANCE == null) {
            INSTANCE = new LazyLoadSingleton();
        }
        return INSTANCE;
    }
}

File: app/src/main/java/app/fynnjason/fivesingletondemo/kotlin/LazyLoadSingleton.kt
package app.fynnjason.fivesingletondemo.kotlin

/**
 * Created by FynnJason on 2017/12/25.
 * 大家熟知的懒加载方式(这里会列出Kotlin的原生写法和翻译Java写法)
 */
class LazyLoadSingleton private constructor(){

    companion object {
        //原生写法
        //LazyThreadSafetyMode.NONE 线程不安全
        val INSTANCE_1 by lazy(LazyThreadSafetyMode.NONE) {
            LazyLoadSingleton()
        }

        //翻译JAVA的写法
        private var INSTANCE_2: LazyLoadSingleton? = null

        fun getInstance(): LazyLoadSingleton {
            if (INSTANCE_2 == null) {
                INSTANCE_2 = LazyLoadSingleton()
            }
            return INSTANCE_2!!
        }
    }
}